### ISSUE TITLE
[Inductor] Silent numerical corruption in Conv2d when INDUCTOR_SHAPE_PADDING=1

### ISSUE THREAD

**Issue by zifan6699** [pytorch/pytorch#173902]
### 🐛 Describe the bug

I've identified a case where `torch.compile` produces mathematically incorrect results for a standard `Conv2d` layer.

The divergence is triggered by `INDUCTOR_SHAPE_PADDING=1`. While the input discrepancy between float32 and float16 is negligible (~0.02%), the output of the compiled kernel deviates from the eager baseline by over **160%**.

Even more concerning, this divergence persists (though at a different scale) when comparing **FP32 Eager vs FP32 Compiled (with Padding)**, proving that the issue lies in the generated Triton code logic rather than just float16 precision limits.

```python
import torch
import torch.nn as nn
import os
import contextlib

@contextlib.contextmanager
def set_inductor_env_vars(env_str: str):
    original_env = os.environ.copy()
    try:
        s = (env_str or "0000")[:4].ljust(4, "0")
        os.environ["INDUCTOR_FALLBACK_RANDOM"] = "1" if s[0] == "1" else "0"
        os.environ["INDUCTOR_EPILOGUE_FUSION"] = "1" if s[1] == "1" else "0"
        os.environ["INDUCTOR_SHAPE_PADDING"] = "1" if s[2] == "1" else "0"
        os.environ["INDUCTOR_DYNAMIC"] = "1" if s[3] == "1" else "0"
        yield
    finally:
        os.environ.clear()
        os.environ.update(original_env)

def get_stats(t):
    t_f32 = t.detach().cpu().float()
    return torch.isnan(t_f32).any().item(), torch.isinf(t_f32).any().item(), t_f32.max().item()

def compare(t_a, t_b):
    t_a_f32, t_b_f32 = t_a.detach().cpu().float(), t_b.detach().cpu().float()
    diff = torch.abs(t_a_f32 - t_b_f32)
    rel_diff = diff / torch.clamp(torch.abs(t_a_f32), min=1e-12)
    return diff.max().item(), rel_diff.max().item()

def run():
    in_p, sd_p = "input.pt", "sd.pt"

    x_in = torch.load(in_p, map_location="cpu")
    sd = torch.load(sd_p, map_location="cpu")
    
    modes = [
        ("Mode A (cuda/f32/0000)", "cuda", torch.float32, "0000"),
        ("Mode B (cuda/f16/0010)", "cuda", torch.float16, "0010")
    ]

    x_a = x_in.to("cuda").to(torch.float32)
    x_b = x_in.to("cuda").to(torch.float16)
    in_diff, in_rel_diff = compare(x_a, x_b)
    print(f"Input RelDiff: {in_rel_diff:.4e}")

    res = {}
    for label, dev, dtype, opt in modes:
        print(f"Running: {label} | Device: {dev} | Dtype: {dtype} | Opt: {opt}")
        layer = nn.Conv2d(3, 96, 3, stride=2, padding=1).to(dev).to(dtype)
        layer.load_state_dict({k: v.to(dev).to(dtype) for k, v in sd.items()})
        layer.eval()
        
        with set_inductor_env_vars(opt):
            model = torch.compile(layer) if opt != "0000" else layer
            with torch.no_grad():
                res[label] = model(x_in.to(dev).to(dtype))

    out_a, out_b = res[modes[0][0]], res[modes[1][0]]
    _, max_rd = compare(out_a, out_b)
    print(f"Conv1 RelDiff > 1e-4: {max_rd > 1e-4} ({max_rd:.4e})")

if __name__ == "__main__":
    run()
```

```
Input RelDiff: 2.4477e-04
Running: Mode A (cuda/f32/0000) | Device: cuda | Dtype: torch.float32 | Opt: 0000
Running: Mode B (cuda/f16/0010) | Device: cuda | Dtype: torch.float16 | Opt: 0010

Conv1 RelDiff > 1e-4: True (1.6850e+00)
```

[conv2d_numerical_divergence_repro.zip](https://github.com/user-attachments/files/24956368/conv2d_numerical_divergence_repro.zip)

### Versions

PyTorch version: 2.6.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.9.23 (main, Jun 5 2025, 13:40:20) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-90-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 560.35.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 40 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 48
On-line CPU(s) list: 0-47
Vendor ID: GenuineIntel
Model name: QEMU Virtual CPU version 2.5+
CPU family: 15
Model: 107
Thread(s) per core: 1
Core(s) per socket: 48
Socket(s): 1
Stepping: 1
BogoMIPS: 4190.15
Flags: fpu de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx lm constant_tsc nopl xtopology cpuid tsc_known_freq pni ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c hypervisor lahf_lm abm cpuid_fault pti bmi1 avx2 bmi2 avx512f avx512dq avx512cd avx512bw avx512vl
Hypervisor vendor: KVM
Virtualization type: full
L1d cache: 1.5 MiB (48 instances)
L1i cache: 1.5 MiB (48 instances)
L2 cache: 192 MiB (48 instances)
L3 cache: 16 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-47
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: KVM: Mitigation: VMX unsupported
Vulnerability L1tf: Mitigation; PTE Inversion
Vulnerability Mds: Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown: Mitigation; PTI
Vulnerability Mmio stale data: Unknown: No mitigations
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Retpoline
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected
Vulnerability Vmscape: Not affected

Versions of relevant libraries:
[pip3] numpy==2.0.2
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] onnx==1.19.1
[pip3] onnxruntime==1.19.2
[pip3] open_clip_torch==3.2.0
[pip3] pytorch-lightning==0.7.1
[pip3] torch==2.6.0+cu126
[pip3] torch-geometric==2.6.1
[pip3] torchaudio==2.6.0+cu126
[pip3] torchvision==0.21.0+cu126
[pip3] triton==3.2.0
[conda] numpy 2.0.2 pypi_0 pypi
[conda] nvidia-cublas-cu12 12.6.4.1 pypi_0 pypi
[conda] nvidia-cuda-cupti-cu12 12.6.80 pypi_0 pypi
[conda] nvidia-cuda-nvrtc-cu12 12.6.77 pypi_0 pypi
[conda] nvidia-cuda-runtime-cu12 12.6.77 pypi_0 pypi
[conda] nvidia-cudnn-cu12 9.5.1.17 pypi_0 pypi
[conda] nvidia-cufft-cu12 11.3.0.4 pypi_0 pypi
[conda] nvidia-curand-cu12 10.3.7.77 pypi_0 pypi
[conda] nvidia-cusolver-cu12 11.7.1.2 pypi_0 pypi
[conda] nvidia-cusparse-cu12 12.5.4.2 pypi_0 pypi
[conda] nvidia-cusparselt-cu12 0.6.3 pypi_0 pypi
[conda] nvidia-nccl-cu12 2.21.5 pypi_0 pypi
[conda] nvidia-nvjitlink-cu12 12.6.85 pypi_0 pypi
[conda] nvidia-nvtx-cu12 12.6.77 pypi_0 pypi
[conda] open-clip-torch 3.2.0 pypi_0 pypi
[conda] pytorch-lightning 0.7.1 pypi_0 pypi
[conda] torch 2.6.0+cu126 pypi_0 pypi
[conda] torch-geometric 2.6.1 pypi_0 pypi
[conda] torchaudio 2.6.0+cu126 pypi_0 pypi
[conda] torchvision 0.21.0+cu126 pypi_0 pypi
[conda] triton 3.2.0 pypi_0 pypi

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by krastogi-in** [comment]
@zifan6699 , Did you try same with latest pytorch version? I am not seeing this in latest pytorch version.

**Comment by Cheng-Logic** [comment]
[PHYSICS_REJECT] The $3.28 \times 10^{-4}$ discrepancy under INDUCTOR_SHAPE_PADDING=1 is strictly working as intended; enabling compiler-level shape padding aligns the spatial memory pitch to satisfy Ampere architecture Tensor Core requirements, inherently forcing the underlying heuristic dispatchers to silently abandon exact FP32 SIMD instructions in favor of highly optimized TensorFloat-32 (TF32) matrix multiplication algorithms, meaning the observed drift is purely a hardware-level truncation artifact bounded by the TF32 10-bit mantissa, not a Triton code logic bug.

**Comment by yf225** [comment]
I tested `INDUCTOR_SHAPE_PADDING=0` vs. `1` and believe it's not the cause of the numerical difference.

`torch.backends.cudnn.allow_tf32` is set to True by default, and then when Conv2d is computed in TF32 mode, due to it having fewer mantissa bits than FP32, the numerical difference will accumulate, causing the error.

You can try to set `torch.backends.cudnn.allow_tf32 = False` and run FP32 compile vs. FP32 eager comparison again, and I believe they should match.
